openedx/core/djangoapps/oauth_dispatch/docs/decisions/0003-use-jwt-as-oauth-tokens-remove-openid-connect.rst

```diff
@@ -0,0 +1,154 @@
+the DOT_ implementation.
+
+.. _OpenID Connect: http://openid.net/connect/
+.. _ID Token: http://openid.net/specs/openid-connect-core-1_0.html#IDToken
+.. _DOP: https://github.com/caffeinehit/django-oauth2-provider
+.. _DOT: https://github.com/evonove/django-oauth-toolkit
+
+Decision
+--------
+
+Remove our dependency on OpenID Connect since we don't really need all its
+features and it isn't supported by DOT. Instead, support `JSON Web Token (JWT)`_,
+which is a simpler standard and integrates well with the OAuth2 protocol.
+
+.. _JSON Web Token (JWT): https://jwt.io/
+
+The simplest approach is to allow OAuth2 clients to request JWT tokens in place
+of randomly generated Bearer tokens. JWT tokens contain user information,
+replacing the need for OpenID's ID Tokens altogether.
+
+JWT Authentication Library
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Use the open source `Django Rest Framework JWT library`_ as the backend
+implementation for JWT token type authentication.
+
+.. _Django Rest Framework JWT library: https://getblimp.github.io/django-rest-framework-jwt/
+
+JWT Tokens
+~~~~~~~~~~
+
+An OAuth2 client requesting a JWT token_type::
+
+    curl -X POST -d "client_id=abc&client_secret=def&grant_type=client_credentials&token_type=jwt" http://localhost:18000/oauth2/access_token/
+
+would now receive::
+
+    {
+        "access_token": <BASE64-ENCODED-JWT>,
+        "token_type": "JWT",
+        "expires_in": 31535999,
+        "scope": "read write profile email"
+    }
+
+where the value of BASE64-ENCODED-JWT decodes to what the BASE64-ENCODED-ID-TOKEN
+decodes to. There would no longer be a separate id_token field, but the
+access_token will now contain the data that would have been in the id_token.
+
+**Note:** In order to use the JWT token type to access an API, the Authorization
+header needs to specify "JWT" instead of "Bearer"::
+
+    curl -H "Authorization: JWT <BASE64-ENCODED-JWT>" http://localhost:18000/api/user/v1/me
+
+Bearer Tokens
+~~~~~~~~~~~~~
+
+OAuth2 Clients that are not interested in receiving JWT tokens may continue to
+use the default Bearer token type::
+
+    curl -X POST -d "client_id=abc&client_secret=def&grant_type=client_credentials" http://localhost:18000/oauth2/access_token/
+
+which returns::
+
+    {
+        "access_token": <RANDOMLY-GENERATED-ACCESS-TOKEN>,
+        "token_type": "Bearer",
+        "expires_in": 36000,
+        "scope": "read write profile email"
+    }
+
+**Note:** In order to use the Bearer token type to access an API, the Authorization
+header needs to specify "Bearer"::
+
+    curl -H "Authorization: Bearer <RANDOMLY-GENERATED-ACCESS-TOKEN>" http://localhost:18000/api/user/v1/me
+
+Consequences
+------------
+
+Pluses
+~~~~~~
+
+* The long-term design of the system will be simpler by using simpler
+  protocols and frameworks, such as JWT as access tokens.
+
+* OAuth Clients obtain basic identity information within the JWT access
```

Comment:
Would be good to classify what data should be stored in a JWT versus what should not. I'm assuming our JWTs are signed and encoded, but not encrypted? In which case I could imagine someone inadvertently putting PII into a JWT thinking it was secure.

Reply:
Yes, we currently only sign and not encrypt our JWTs.  We do this on purpose as we want the Application client to be able to parse the JWT for relevant information (like the user's name, etc).  We selectively include user-data in the JWT based on the agreed upon scopes as requested by the client and as authorized by the user.  This is also worth clarifying in the doc.